### ISSUE TITLE
resolved references/selectors are immediately removed when they are arrays or maps in v1.14.x

### ISSUE THREAD

**Issue by haarchri** [crossplane/crossplane#5055]
### What happened?
see for reference: https://github.com/upbound/provider-aws/issues/975

The issue is that, for instance, the EKS cluster managed resource successfully resolves subnetIds. However, we observe that subnetIds and subnetIdRefs are consistently removed every 30 seconds, causing the fields to fluctuate. I suspect that Crossplane removes the resolved fields, and this issue specifically affects arrays and maps or situations with multiple resolves.

![image](https://github.com/crossplane/crossplane/assets/21083497/0ce6430e-251f-4767-9ca4-433584e4ce37)

```
Vpc Config:
   Endpoint Private Access: true
   Endpoint Public Access:  true
   Subnet Id Refs:
    Name: configuration-aws-eks-nwlbx
    Name: configuration-aws-eks-6tjj7
   Subnet Id Selector:
    Match Labels:
     Access:                    public
     networks.aws.platform.upbound.io/network-id: configuration-aws-eks
   Subnet Ids:
    subnet-088e1fe579611d91f
    subnet-0b5b8e8e15c636284
```
vs.

```
    Vpc Config:
      Endpoint Private Access:  true
      Endpoint Public Access:   true
      Subnet Id Selector:
        Match Labels:
          Access:                                       public
          networks.aws.platform.upbound.io/network-id:  configuration-aws-eks
```

### How can we reproduce it?

clone the following repository and run uptest with `make e2e` https://github.com/upbound/configuration-aws-eks

full manifest:

```
kubectl get cluster.eks.aws.upbound.io/configuration-aws-eks-knvvh  -o yaml --show-managed-fields
apiVersion: eks.aws.upbound.io/v1beta1
kind: Cluster
metadata:
  annotations:
    crossplane.io/composition-resource-name: kubernetesCluster
    crossplane.io/external-name: configuration-aws-eks-knvvh
    upjet.upbound.io/test: "true"
  creationTimestamp: "2023-11-22T22:00:28Z"
  generateName: configuration-aws-eks-
  generation: 94
  labels:
    crossplane.io/claim-name: ""
    crossplane.io/claim-namespace: ""
    crossplane.io/composite: configuration-aws-eks
  managedFields:
  - apiVersion: eks.aws.upbound.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:upjet.upbound.io/test: {}
    manager: kubectl-annotate
    operation: Update
    time: "2023-11-22T22:00:40Z"
  - apiVersion: eks.aws.upbound.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        .: {}
        f:atProvider: {}
        f:conditions:
          .: {}
          k:{"type":"Synced"}:
            .: {}
            f:lastTransitionTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
    manager: provider
    operation: Update
    subresource: status
    time: "2023-11-22T22:00:40Z"
  - apiVersion: eks.aws.upbound.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:crossplane.io/composition-resource-name: {}
        f:generateName: {}
        f:labels:
          .: {}
          f:crossplane.io/claim-name: {}
          f:crossplane.io/claim-namespace: {}
          f:crossplane.io/composite: {}
        f:ownerReferences:
          .: {}
          k:{"uid":"d7b58e1a-7494-46d9-ac6e-517ab63a3acf"}: {}
      f:spec:
        .: {}
        f:deletionPolicy: {}
        f:forProvider:
          .: {}
          f:region: {}
          f:roleArnSelector:
            .: {}
            f:matchControllerRef: {}
            f:matchLabels:
              .: {}
              f:role: {}
          f:version: {}
          f:vpcConfig: {}
        f:managementPolicies: {}
        f:providerConfigRef:
          .: {}
          f:name: {}
    manager: crossplane
    operation: Update
    time: "2023-11-22T22:04:08Z"
  - apiVersion: eks.aws.upbound.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:crossplane.io/external-name: {}
      f:spec:
        f:forProvider:
          f:roleArn: {}
          f:roleArnRef:
            .: {}
            f:name: {}
          f:tags:
            .: {}
            f:crossplane-kind: {}
            f:crossplane-name: {}
            f:crossplane-providerconfig: {}
        f:initProvider: {}
    manager: provider
    operation: Update
    time: "2023-11-22T22:04:08Z"
  name: configuration-aws-eks-knvvh
  ownerReferences:
  - apiVersion: aws.platform.upbound.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: XEKS
    name: configuration-aws-eks
    uid: d7b58e1a-7494-46d9-ac6e-517ab63a3acf
  resourceVersion: "4052"
  uid: b8e5e1cb-b165-4a1b-96cb-7cf451f47476
spec:
  deletionPolicy: Delete
  forProvider:
    region: us-west-2
    roleArn: arn:aws:iam::609897127049:role/configuration-aws-eks-lgbkm
    roleArnRef:
      name: configuration-aws-eks-lgbkm
    roleArnSelector:
      matchControllerRef: true
      matchLabels:
        role: controlplane
    tags:
      crossplane-kind: cluster.eks.aws.upbound.io
      crossplane-name: configuration-aws-eks-knvvh
      crossplane-providerconfig: default
    version: "1.27"
    vpcConfig:
    - endpointPrivateAccess: true
      endpointPublicAccess: true
      subnetIdSelector:
        matchLabels:
          access: public
          networks.aws.platform.upbound.io/network-id: configuration-aws-eks
  initProvider: {}
  managementPolicies:
  - '*'
  providerConfigRef:
    name: default
status:
  atProvider: {}
  conditions:
  - lastTransitionTime: "2023-11-22T22:00:40Z"
    message: 'cannot resolve references: mg.Spec.ForProvider.VPCConfig[i3].SubnetIds:
      referenced field was empty (referenced resource may not yet be ready)'
    reason: ReconcileError
    status: "False"
    type: Synced
```

```
kubectl get cluster.eks.aws.upbound.io/configuration-aws-eks-knvvh  -o yaml --show-managed-fields
apiVersion: eks.aws.upbound.io/v1beta1
kind: Cluster
metadata:
  annotations:
    crossplane.io/composition-resource-name: kubernetesCluster
    crossplane.io/external-name: configuration-aws-eks-knvvh
    upjet.upbound.io/test: "true"
  creationTimestamp: "2023-11-22T22:00:28Z"
  generateName: configuration-aws-eks-
  generation: 155
  labels:
    crossplane.io/claim-name: ""
    crossplane.io/claim-namespace: ""
    crossplane.io/composite: configuration-aws-eks
  managedFields:
  - apiVersion: eks.aws.upbound.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:upjet.upbound.io/test: {}
    manager: kubectl-annotate
    operation: Update
    time: "2023-11-22T22:00:40Z"
  - apiVersion: eks.aws.upbound.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        .: {}
        f:atProvider: {}
        f:conditions:
          .: {}
          k:{"type":"Synced"}:
            .: {}
            f:lastTransitionTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
    manager: provider
    operation: Update
    subresource: status
    time: "2023-11-22T22:00:40Z"
  - apiVersion: eks.aws.upbound.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:crossplane.io/composition-resource-name: {}
        f:generateName: {}
        f:labels:
          .: {}
          f:crossplane.io/claim-name: {}
          f:crossplane.io/claim-namespace: {}
          f:crossplane.io/composite: {}
        f:ownerReferences:
          .: {}
          k:{"uid":"d7b58e1a-7494-46d9-ac6e-517ab63a3acf"}: {}
      f:spec:
        .: {}
        f:deletionPolicy: {}
        f:forProvider:
          .: {}
          f:region: {}
          f:roleArnSelector:
            .: {}
            f:matchControllerRef: {}
            f:matchLabels:
              .: {}
              f:role: {}
          f:version: {}
        f:managementPolicies: {}
        f:providerConfigRef:
          .: {}
          f:name: {}
    manager: crossplane
    operation: Update
    time: "2023-11-22T22:04:45Z"
  - apiVersion: eks.aws.upbound.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:crossplane.io/external-name: {}
      f:spec:
        f:forProvider:
          f:roleArn: {}
          f:roleArnRef:
            .: {}
            f:name: {}
          f:tags:
            .: {}
            f:crossplane-kind: {}
            f:crossplane-name: {}
            f:crossplane-providerconfig: {}
          f:vpcConfig: {}
        f:initProvider: {}
    manager: provider
    operation: Update
    time: "2023-11-22T22:04:46Z"
  name: configuration-aws-eks-knvvh
  ownerReferences:
  - apiVersion: aws.platform.upbound.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: XEKS
    name: configuration-aws-eks
    uid: d7b58e1a-7494-46d9-ac6e-517ab63a3acf
  resourceVersion: "4335"
  uid: b8e5e1cb-b165-4a1b-96cb-7cf451f47476
spec:
  deletionPolicy: Delete
  forProvider:
    region: us-west-2
    roleArn: arn:aws:iam::609897127049:role/configuration-aws-eks-lgbkm
    roleArnRef:
      name: configuration-aws-eks-lgbkm
    roleArnSelector:
      matchControllerRef: true
      matchLabels:
        role: controlplane
    tags:
      crossplane-kind: cluster.eks.aws.upbound.io
      crossplane-name: configuration-aws-eks-knvvh
      crossplane-providerconfig: default
    version: "1.27"
    vpcConfig:
    - endpointPrivateAccess: true
      endpointPublicAccess: true
      subnetIdRefs:
      - name: configuration-aws-eks-cqgc5
      - name: configuration-aws-eks-w6mwj
      subnetIdSelector:
        matchLabels:
          access: public
          networks.aws.platform.upbound.io/network-id: configuration-aws-eks
      subnetIds:
      - subnet-0b4b45026d7cb1b6b
      - subnet-0d5a5e33bab277d90
  initProvider: {}
  managementPolicies:
  - '*'
  providerConfigRef:
    name: default
status:
  atProvider: {}
  conditions:
  - lastTransitionTime: "2023-11-22T22:00:40Z"
    message: 'cannot resolve references: mg.Spec.ForProvider.VPCConfig[i3].SubnetIds:
      referenced field was empty (referenced resource may not yet be ready)'
    reason: ReconcileError
    status: "False"
    type: Synced
```

### What environment did it happen in?
Crossplane version:  v1.14.2

**Comment by haarchri** [comment]
To provide clarification, the same issue occurs with both traditional patch & transform and patch & transform functions. I experimented with v1.13.x, and it seems like I observe a similar "issue". However, in that version, Crossplane may not be quick enough to remove the fields before the managed resources are created.

**Comment by jbw976** [comment]
From what @haarchri has described so far here, this looks fairly concerning and worth more investigation and a potential fix to backport to a patch release for v1.14.

**Comment by ytsarev** [comment]
It looks like a duplicate of https://github.com/crossplane/crossplane/issues/3335 . 

Because of this situation, we had to make a workaround with explicit `subneIds` propagation through XR status with bi-directional patching in https://github.com/upbound/platform-ref-aws/pull/69 , specifically https://github.com/upbound/platform-ref-aws/pull/69/commits/dd43bfa67637ed114c92889cd289e6e6fddeee78 and https://github.com/upbound/platform-ref-aws/pull/69/commits/e0f4241be8fe8fd194c9369d1ad5fb599e6b6c61

Thanks for raising attention to this important unsolved problem 👍 We had this issue long before 1.14 though. 
CC @jeanduplessis

**Comment by sttts** [comment]
The best solution, other than complicated work-arounds is to fix the CRD schema in the provider. Something along these lines:
```yaml
type: array
x-kubernetes-list-type: Map
x-kubernetes-list-key: id
items:
  properties:
    id:
      type: string
      default: "singleton"
```
for the array.

**Comment by haarchri** [comment]
x-ref https://github.com/crossplane-contrib/function-patch-and-transform/issues/50 https://github.com/crossplane-contrib/provider-helm/issues/204

**Comment by ulucinar** [comment]
We've rolled out SSA patching while updating the resolved `subnetId` references in the `spec.forProvider.vpcConfig[]` object list in order to address https://github.com/upbound/provider-aws/issues/975 with [`upbound/provider-aws@v0.47.0`](https://github.com/upbound/provider-aws/releases/tag/v0.47.0). Also validated the fix with the [P&T composition function v0.2.1](https://marketplace.upbound.io/functions/upbound/function-patch-and-transform/v0.2.1) with the composition described [here](https://github.com/upbound/provider-aws/pull/1046#issue-2057334633).

**Comment by jeanduplessis** [comment]
Closing this issue in favor of it needing to be addressed in the providers, rather than crossplane/crossplane.
